www/js/controllers.js

```diff
@@ -48,55 +47,94 @@ app.controller('HistoryController',function($rootScope, $scope, Auth, UserFinish
-        $scope.isNewTask = item == undefined;
-        $scope.task = item ? item : {} ;
-        $scope.modal.show();
+        } else {
+            $scope.dueTime = task.dueDate - (new Date().getTime());
+            var timer = $interval( function(){
+                if($scope.dueTime == 0){
+                    $interval.cancel(timer);
+                }
+                $scope.dueTime = $scope.dueTime - 1000;
+            }, 1000);
+        }
     };
-
-    $scope.addTask = function (task) {
-        if (task.message.trim() == '') {
+    if ($scope.isNewTask) {
+        var createdDate = new  Date();
+        $scope.task = {
+            ownerId: ownerId,
+            ownerName: getDisplayNameFromAuth(Auth.$getAuth()),
+            createdAt: createdDate.getTime(),
+            dueDate : 0
+        }
+    } else {
+        $scope.task = UserTask(ownerId, $stateParams.taskId);
+        $scope.setTimer($scope.task);
+    }
+    $scope.saveTask = function (task) {
+        function failureCallback(error) {
             // TODO Show error.
+        }
+        if ( task == undefined || task.message.trim() == '') {
+            failureCallback();
             return;
         }
-        $scope.items.$add({message: task.message}).then(function(ref) {
-            console.log(">> Successfully saved a message ", ref.key(), ref.$id);
-            task.message = '';
-            $scope.closeNewTask();
-        }).catch(function(error) {
-            // TODO show error on the modal itself.
-        });
+        function successCallback(ref) {
+            $ionicHistory.goBack();
+        }
+        if (!task.$id) {
+            var tasks = UserTasks(ownerId);
+            tasks.$add(task).then(successCallback).catch(failureCallback);
+        } else {
+            task.$save().then(successCallback).catch(failureCallback);
+        }
     };
-
-    $scope.updateTask = function (item) {
-        $scope.items.$save(item).then(function(ref) {
-            if (ref.key() == item.$id) {
-                console.log(">> Successfully updated ", ref.key());
-                $scope.closeNewTask();
-                $ionicListDelegate.closeOptionButtons();
-            }
-        }).catch(function(error) {
-        });
+    $scope.dateReminder =function(task) {
+            $cordovaDatePicker.show(getDateOptions('date')).then(function (date) {
+                var reminderDate = moment(date).format("ddd MMM D YYYY");
+                $scope.timeReminder(task, reminderDate);
+            });
     };
+    $scope.timeReminder = function(task, reminderDate) {
+            $cordovaDatePicker.show(getDateOptions('time')).then(function (date) {
+                var reminderTime = moment(date).format("HH:mm:ss");
+                task.dueDate = moment(reminderDate + " " +reminderTime).toDate().getTime();
+                task.$save().then(function(){
+                }).catch(function(){
+                    $scope.setTimer(task);
```

Comment:
$scope.setTimer should get called irrespective of success/failure. (`then` function is success. `catch` function is failure).   They are similar to `try/catch` block in java. 

also, if it is a new task, you shouldnt try to save it. you should wait for the user to save it.